### PR TITLE
Add subsection on multiple observation model comparison

### PR DESCRIPTION
**Pull request by OriolAbril**
As discussed in #14, this PR adds a new subsection to model comparison section. It is based on the rugby model but applied to premier league data. I still have some doubts about the best way to organize the notebook and what to add to make it clear and accessible. Here are some of them, but feel free to add extra suggestions:

1. How should the explanation on the different questions be tackled? There will surely be the ArviZ approach of concatenating/adding pointwise log likelihood values (this is already in [the notebook](https://github.com/OriolAbril/Exploratory-Analysis-of-Bayesian-Models/blob/multi_obs_ic/content/Section_04/Multiple_observations.ipynb)), what else should accompany these calculations:
   * Conceptual explanation on mimicking cross validation leaving one match/goal recording/team out
   * Mathematical reparametrization of the log likelihood in terms of the desired "unit observation". See [discourse clarification](https://discourse.pymc.io/t/calculating-waic-for-models-with-multiple-likelihood-functions/4834/7?u=oriolabril)
   * Auxiliar PyMC model implementing the reparametrization above. See [this notebook](https://github.com/OriolAbril/calaix_de_sastre/blob/master/premier-hierarchical-model/premier-analytics.ipynb)
1. I think this should be only about IC calculation and refer to model comparison subsection once IC calculation is obtained, but I'm not 100% sure about this. Would it make sense to compare to negative binomial for example?
1. I need help with terminology :sweat_smile: leave one team out and leave one match out I think are ok, but leave _half match_ out? I used observation in discourse but I think it was a bad idea

### CODE REVIEW

File: content/Section_04/Multiple_observations.ipynb [hunk 1]
```diff
@@ -0,0 +1,997 @@
+{
```
Comment (AlexAndorra) [inline content/Section_04/Multiple_observations.ipynb:1]: <p>Maybe do the imports alphabetically? And absolute imports before relative ones (although not relevant here)?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (AlexAndorra) [inline content/Section_04/Multiple_observations.ipynb:1]: <p>If you're running PyMC master or &gt;= 3.9, you just do <code>home_team = pm.Data("home_team", home_team_idxs)</code>  ;)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (AlexAndorra) [inline content/Section_04/Multiple_observations.ipynb:1]: <p>Can you add a last cell with the <code>watermark</code> <a href="https://github.com/rasbt/watermark" target="_blank">package</a> to help with reproducibility:</p><pre class="ql-syntax" spellcheck="false">%load_ext watermark
%watermark -n -u -v -iv -w
</pre><p>Let's also black-format with <a href="https://github.com/dfm/black_nbconvert" target="_blank">black_nbconvert</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

**Comment by AlexAndorra** [comment]
Thanks Oriol, this is a very good start! Regarding your questions:

1. I agree with your first points -- this should really be in the NB, on the model on the explanation you wrote in the rugby NB for Discourse. More generally, a thorough conceptual explanation of the model is always welcome and will help with understanding the subsequent CVs. Your two other points seem less necessary to inclued in the NB to me -- or maybe just an explanation of the reparametrization of the log likelihood in terms of the desired unit observation, but maybe not the implementation of the models.

2. I agree with you. Neg Binomial sounds good, but maybe also a hierarchical model? I think there are interesting clusters of teams here.

3. I understood it was leave one match / goal / team out; I didn't see leave half a match out 🤔 More generally, what would be useful is explaining the concept of "leave one X out" -- why would we need to do that?

Finally, I think this would benefit from prior and posterior predictive checks, as well as posterior plots and trace plots. I'm very happy to contribute these parts, although I don't think I'll have time this week 😬

**Comment by OriolAbril** [comment]
I will get to watermark and sorting imports eventually, for now I prefer to focus on the content. i would like to eventually run it on pymc3.9 and latest ArviZ

1. I think that the discussion on the exchangeability condition should be present here to better serve the goal of the repo, I am not sure this is possible without the mathematical description of the likelihood nor the reparametrization example.

1. I think this should be focused on information criteria and how to use them with multiple observations and hierarchical models, adding one extra model to perform the comparison all the way could serve that goal, but I don't want to get lost in a full model comparison and iterative improvement of the model.

1. goal/observation/"half match" could all be used but I think that all of them can still be confusing :/

> Finally, I think this would benefit from prior and posterior predictive checks, as well as posterior plots and trace plots. I'm very happy to contribute these parts, although I don't think I'll have time this week 

As I said above, not sure this fits the goal of the section, there will be other sections on ppc, trace plots... We could look into making this a complete case study (and even extend the data to several years or several countries at the same time) but that would be a whole different story, I think this should be laser focused on ic plus multiple observations (this is why I am not even sure about model comparison, there already is a subsection on model comparison and I don't want to be too repetitive)

**Comment by AlexAndorra** [comment]
> I think that the discussion on the exchangeability condition should be present here to better serve the goal of the repo, I am not sure this is possible without the mathematical description of the likelihood nor the reparametrization example.

I always like when there is more details to explain a method than less, but I'm trying to find the "least long" explanation possible: the discussion on exchangeability is necessary, and if you think the mathematical description of the likelihood is too to bring the point home, then let's add it too. And I guess while writing this we will see whether the reparametrization examples should be added too -- to be clear, I think they are of value, but they take a lot of place; it's a trade-off.

> I think this should be focused on information criteria and how to use them with multiple observations and hierarchical models, adding one extra model to perform the comparison all the way could serve that goal, but I don't want to get lost in a full model comparison and iterative improvement of the model.

I see what you mean. A good idea could be to use this model and show how to improve it across various NBs of this repo, each NB being focused on one aspect of the iterative process. This would constitute a complete case-study, as you say, and I think this model is interesting because there are natural hierarchical structures (plus, it's _not_ the Iris or Titanic datasets 😜 ).
I understand the need to be focused on IC, but I think we shouldn't give people the impression that you can use IC in a vacuum -- you use them in a specific part of the modeling worfklow, so putting them in context seems important to me.

> goal/observation/"half match" could all be used but I think that all of them can still be confusing

The last are definitely confusing, but the first one is quite clear to me: what's the expected predictive of the models if we were trying to predict which team will score the next goal? But maybe I misunderstood what "leave one goal out" means 😆

File: content/Section_04/Multiple_observations.ipynb (hunk 1 above)
Comment (canyon289) [inline content/Section_04/Multiple_observations.ipynb:1]: <p>More of a question for myself. Is there somewhere that explains what the differences are between pointwise and "not pointwise" easily?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

**Comment by canyon289** [comment]
To your question about splitting this out.

> I think this should be only about IC calculation and refer to model comparison subsection once IC calculation is obtained, but I'm not 100% sure about this. Would it make sense to compare to negative binomial for example?

I would agree that explaining IC calculation split from comparison seems to make sense? As far as the code goes its a great example of how to do model checks with multiple likelihoods.

As far as soccer/football goes I don't know what a half match is, so I lack the domain knowledge to know whats going on.

I like the direction this is going, am interested to see the text that explains what is going on!

File: content/Section_04/Multiple_observations.ipynb (hunk 1 above)
Comment (OriolAbril) [reply content/Section_04/Multiple_observations.ipynb]: I am not sure I understand the question. `waic` and `loo` are calculated from pointwise log likelihood values in the sense that some calculations are done with the likelihood corresponding to each observation, then these intermediate values (pointwise ic values) are aggregated to caclulate loo/waic elpd value. By default ArviZ does not return the pointwise ic values (and therefore looses some info like khat values), setting this here changes the default so loo returns pointwise values too and all prints show khat summary. The elpd value is the same.

**Comment by OriolAbril** [comment]
I have extended the content with some math explanation and alternative implementations, now there is actually content to review, I am still not sure about what to include and what to exclude but most of the content was already written I basically had to gather the pieces and put them together. I think it will be clear now, most of the work should go into making this both clear and concise, it is still a little caothic.

**Comment by AlexAndorra** [comment]
Thanks Oriol, just skimmed through it and it looks awesome! Here is a first batch of comments.
I stopped at "Information criterion calculation" and will come back to it later today 😉

File: content/Section_04/Multiple_observations.ipynb (hunk 1 above)
Comment (canyon289) [reply content/Section_04/Multiple_observations.ipynb]: Ok I think get the difference between pointwise now and not but need to get learned up. No issue on your part

**Comment by OriolAbril** [comment]
<p>Good point, I'll change this. I will probably agree with any proposal to reduce the number of times <em>observation</em> is used, I think it is ambiguous in this context. </p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by OriolAbril** [comment]
<p>In the league there are 20 teams which means that there are 38 match days (they play twice against each other except themselves) and there are 10 matches every day. <code>g</code>  index indicates the match id (not the day) out of <code>n=380</code>  matches . Therefore y_g,h are scalars. In fact we have no info at all about the match days in this model.</p><p> </p><p> </p><p> </p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by OriolAbril** [comment]
<p>I'd defer readers interested in that to go to the source of the model (I have in mind to add the sources on top like in the pymc3 example notebook). </p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by aloctavodia** [comment]
I still not have the time to review this, but as a general comment about this repository motivated by question 1. I think each notebook should focus on a single topic (and we can go as granular as we want), if we want to show a more complete "bayesian worflow" we should have a dedicated notebook (or notebooks) to do that. And we should try to discuss as much theory as possible (always keeping in mind the theoretical elements that are useful for the applications). As not every potential user of this repo will be interested in "going to deep" on the theoretical side we may have a "in depth section" per notebook. in fact a few of the already available notebook have it. I am not saying we must follow this pattern in every notebook, but we can use if necessary or desired. Another (not mutually exclusive) option is to have a few notebooks more theoretical and another more practical.

**Comment by AlexAndorra** [comment]
<p>Ok, much clearer, thanks!</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by AlexAndorra** [comment]
<p>Ow, there is a source for this model?? I'm interested!</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by AlexAndorra** [comment]
Thanks @OriolAbril, this is really nice now! The explanation are very clear and the examples easy to follow 👏 
I added some comments and suggested some changes below. Happy to discuss with you and implement them if you want / don't have time 🙂

**Comment by OriolAbril** [comment]
<p>I would say one goal of the notebook is understanding why ArviZ fails with multiple likelihoods, however the error may be read by users who have not read the notebook, it is in this case that I am not sure the error provides any useful info.</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by OriolAbril** [comment]
<p>yeah, I'll change that. I had both cases, but the only difference is changing the home by away, so I figured it was not worth it and I don't know why I erased the away one :shrug:</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by OriolAbril** [comment]
<p>I am still not sure about adding the alternative implementations, I would prefer to keep the notebook backend agnostic (I think it will be possible once the updated rugby data is available in ArviZ). I already had them so I decided to include them in case they could help any of you clarify some doubts. </p><p> </p><p>Moreover, I think it will be better to use some diagrams to highlight the observations we are interested in. I'll add them whenever I have time. </p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by OriolAbril** [comment]
<p>I used the code in pymc3 rugby example which in turn is based in premier example http://danielweitzenfeld.github.io/passtheroc/blog/2014/10/28/bayes-premier-league/</p><p> </p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by OriolAbril** [comment]
<p>I agree if we eventually add the code it should probably have some explanation or link to a description of pm.Potential, as I said above though, I am not sure about this being inside the scope of the notebook.</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by OriolAbril** [comment]
<p>I think the second bullet point is a little confusing/imprecise: what we are doing here is assessing the predictive accuracy of predicting the outcome of the whole match. If our ultimate goal is the one above (going to matches that have the most probability of ending 3-3 or 4-4) this strategy would be the one to use in order to get the predictive accuracy of the desired "observation" (if we had several models, we would compare them based on these loo values instead of away team predictive accuracy). But there are other cases where the interest lies in the whole outcome of the match, the 3-3 case is only one example. Other silly examples could be: wanting to go to the dullest match because you don't care much about the match and want to be able to talk during it (maybe you are bringing your significant other); having some kind of fetish for matches that end up 4-1; predicting the whole outcome of matches to bet on them... </p><p> </p><p>In the same betting example,  you may realize there is more money in guessing the goals of the away team than the goals of the home team and therefore you'd compare models with the away goals metric (like the supporters with low budget). I don't want to use betting examples though.</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by AlexAndorra** [comment]
<p>Ow ok, outside of this NB context you mean. Maybe we can add a link to this NB in the error message, once the NB is merged?</p><p> </p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by AlexAndorra** [comment]
<p>Diagrams are a great idea! And it's true that if we add diagrams, then the other implementations of the model are less necessary (although interesting to keep somewhere). That'd be great if we could hide/show cells in Jupyter NBs (like with the new <code>InferenceData</code> HTML repr). That way, we could hide the cells with the other models by default and users could look at them if needed.</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by AlexAndorra** [comment]
<p>To be clear, I had already understood what you explicited above with the material already present in the NB. I like your point about changing the precise example, instead of taking the same as in the bullet point. That way, people will understand that 3-3 or 4-4 is just an example.</p><p>Something like:</p><p>"Here, the utility we're trying to maximize, as slightly nerdy football fans, is the <em>pairs of goals</em> -- matches that best fit our football tastes could be those that end up 4-4, or those with the lowest number of goals because we don't care much about the game and want to be able to talk during it, or those that end up with a precise score, like 6-2 (because we also love tennis)."</p>

---

 <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by OriolAbril** [comment]
Here is a preview of the kind of diagrams I had in mind. I would include only the picture and not the code to generate it (i would then probably make a blog post with the code and diagram generation). (note the diagrams are for rugby dataset, not premier league)

image

image

image


source code: https://nbviewer.jupyter.org/github/OriolAbril/oriol_unraveled/blob/multi_lik/_notebooks/2020-06-21-multi-likelihood-diagrams.ipynb

**Comment by AlexAndorra** [comment]
Good idea, that looks nice! And indeed, releasing the code in a subsequent blog post seems appropriate -- showing how to display this figure isn't the core of this NB. We'll have to explain how to interpret the diagram though.

**Comment by OriolAbril** [comment]
I think I got the right skeleton now, only big structural change could be using `rugby` dataset instead to delete all backend specific info. Let me know what you think.

File: content/Section_04/Multiple_likelihoods.ipynb [hunk 2]
```diff
@@ -0,0 +1,3438 @@
+{
```
Comment (canyon289) [inline content/Section_04/Multiple_likelihoods.ipynb:1]: <p>Small typo: straighforward  is mispelled</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/arviz-devs/Exploratory-Analysis-of-Bayesian-Models/pull/27/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

**Comment by aloctavodia** [comment]
@OriolAbril I you are OK, I would like to merge this, and then work on adapting it to PyMC 5 and integrate it with the rest of the chapters.

**Comment by aloctavodia** [comment]
I have not used it yet, but Quarto has a freeze feature https://quarto.org/docs/projects/code-execution.html#freeze 
The other option (more DIY style) is to run the models once (or manually when needed) and write the code for the model in markdown cells instead of code cells, so they don't execute. We will figure it out. For the moment my main goal is to tidy up stuff.

**Pull request merged**